fix asan failure for fold_conv_bn when conv is quantized

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42041 optimize_for_mobile: bring packed params to root module
* **#42461 fix asan failure for fold_conv_bn when conv is quantized**

Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

vkuzo committed:
[demo only] asan failure for module freezing

Summary:

This is a test case which fails with ASAN on at the module freezing
step.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on test/test_mobile_optimizer.py:
```diff
@@ -207,5 +208,45 @@ def get_lint_count_by_type(lint_type, module_lint_List):
+            model(torch.randn(4, 1, 4, 4))
+            torch.quantization.convert(model, inplace=True)
+            model = torch.jit.script(model)
+            model_optim = optimize_for_mobile(model)
```
fails here

vkuzo committed:
Update on "[demo only] asan failure for module freezing"


Summary:

This is a test case which fails with ASAN on at the module freezing
step.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on test/test_mobile_optimizer.py:
```diff
@@ -207,5 +208,46 @@ def get_lint_count_by_type(lint_type, module_lint_List):
+        torch.quantization.convert(model, inplace=True)
+        model = torch.jit.script(model)
+        # this line should not have ASAN failures
+        model_optim = optimize_for_mobile(model)
```
if anyone has suggestions on something better to assert here (currently just a no-op asserting that no exceptions are raised), I'm flexible

jerryzh168 replied:
I think it's fine to omit the asserts if we just want to make sure it runs

vkuzo commented on torch/csrc/jit/passes/fold_conv_bn.cpp:
```diff
@@ -59,13 +59,21 @@ void replaceConvBiasWithGetAttr(Module& module) {
-           pattern_name))) {
+           pattern_name));
+  // we need to treat quantized modules in a special way
+  bool real_typename_contains_quantized =
```
if there is a better way to check for this which wouldn't catch people naming their floating point modules ".*quantized.*", I'm flexible

jerryzh168 replied:
can we do this: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16is_conv2d_moduleERKNS0_5MatchERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPNS0_5ValueESt4hashISA_ESt8equal_11899141

jerryzh168 replied:
we can add bias only when the module is conv{1,2,3}d

vkuzo committed:
improve test on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

vkuzo committed:
clang on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

vkuzo committed:
better conv type check on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

vkuzo committed:
clang on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

vkuzo committed:
add qnnpack check on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good, thanks!

vkuzo committed:
fix tests and add convrelu on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

vkuzo committed:
rebase on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

vkuzo committed:
rebase on "fix asan failure for fold_conv_bn when conv is quantized"


Summary:

1. Adding a test case which fails on master with ASAN on at the module freezing step.
2. Adding a fix for the test case.  The problem was that fold_conv_bn attempted to add bias to quantized conv instances.

Test Plan:

```
USE_ASAN=1 USE_CUDA=0 python setup.py develop
LD_PRELOAD=/usr/lib64/libasan.so.4 python test/test_mobile_optimizer.py TestOptimizer.test_optimize_for_mobile_asan

// output tail: https://gist.github.com/vkuzo/7a0018b9e10ffe64dab0ac7381479f23 (before the fix)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960843](https://our.internmc.facebook.com/intern/diff/D22960843)

[ghstack-poisoned]

vkuzo commented:
resubmitted as https://github.com/pytorch/pytorch/pull/42739 due to a ghstack snafu
